pkg/webhook/kafkacluster_validator.go

```diff
@@ -0,0 +1,79 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	banzaicloudv1beta1 "github.com/banzaicloud/koperator/api/v1beta1"
+)
+
+func (s *webhookServer) validateKafkaCluster(kafkaClusterNew *banzaicloudv1beta1.KafkaCluster) *admissionv1.AdmissionResponse {
+	ctx := context.Background()
+	// get the Old kafkaCluster CR
+	kafkaClusterSpecOld := banzaicloudv1beta1.KafkaCluster{}
+	err := s.client.Get(ctx, client.ObjectKey{Name: kafkaClusterNew.GetName(), Namespace: kafkaClusterNew.GetNamespace()}, &kafkaClusterSpecOld)
+	if err != nil {
+		// New kafkaCluster has been added thus no need to check storage removal
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		log.Error(err, "couldn't get KafkaCluster custom resource")
+		return notAllowed("API failure while retrieving KafkaCluster CR, please try again", metav1.StatusReasonServiceUnavailable)
+	}
+
+	res := checkBrokerStorageRemoval(&kafkaClusterSpecOld.Spec, &kafkaClusterNew.Spec)
+	if res != nil {
+		return res
+	}
+
+	// everything looks a-okay
+	return &admissionv1.AdmissionResponse{
+		Allowed: true,
+	}
+}
+
+// checkBrokerStorageRemoval checking is there any broker storage which has been removed. If yes, admission will be refused
+func checkBrokerStorageRemoval(kafkaClusterSpecOld, kafkaClusterSpecNew *banzaicloudv1beta1.KafkaClusterSpec) *admissionv1.AdmissionResponse {
+	for _, brokerOld := range kafkaClusterSpecOld.Brokers {
+		for _, brokerNew := range kafkaClusterSpecNew.Brokers {
+			if brokerOld.Id == brokerNew.Id {
+				brokerConfigsOld, _ := brokerOld.GetBrokerConfig(*kafkaClusterSpecOld)
+				brokerConfigsNew, _ := brokerNew.GetBrokerConfig(*kafkaClusterSpecNew)
+				for _, storageConfigOld := range brokerConfigsOld.StorageConfigs {
+					isStorageFound := false
+
+					for _, storageConfigNew := range brokerConfigsNew.StorageConfigs {
+						if storageConfigOld.MountPath == storageConfigNew.MountPath {
+							isStorageFound = true
+							break
+						}
+					}
+					if !isStorageFound {
+						log.Info(fmt.Sprintf("Not allowed to remove broker storage with mountPath: %s from brokerID: %v", storageConfigOld.MountPath, brokerOld.Id))
+						return notAllowed(fmt.Sprintf("Removing storage from a runnng broker is not supported! (mounthPath: %s, brokerID: %v)", storageConfigOld.MountPath, brokerOld.Id), metav1.StatusReasonInvalid)
+					}
+				}
+			}
+		}
+	}
```

Comment:
Im not sure it would be faster. hash function also need time. searching in an array sometimes faster when the array is not so big. There is more chance for cache hit because locality. I will make a benchmark and I will write the result here then. But Im almost sure if we don't have 10000 broker the iteration is faster then the map.

Reply:
I have made tests:
With 10 brokers:
BenchmarkCheckBrokerStorageRemoval-12             336151             17843 ns/op            7709 B/op        181 allocs/op
BenchmarkCheckBrokerStorageRemovalMap-12          315955             20312 ns/op            7709 B/op        181 allocs/op
with 200 brokers:
BenchmarkCheckBrokerStorageRemoval-12               1857           3301253 ns/op         1534065 B/op      36213 allocs/op
BenchmarkCheckBrokerStorageRemovalMap-12            1784           3638668 ns/op         1566985 B/op      36224 allocs/op
With 1000 brokers:
BenchmarkCheckBrokerStorageRemoval-12                235          24051552 ns/op         7701831 B/op     181791 allocs/op
BenchmarkCheckBrokerStorageRemovalMap-12             273          21840247 ns/op         7959142 B/op     181814 allocs/op
With 20000 brokers:
BenchmarkCheckBrokerStorageRemoval-12                  6        1067613391 ns/op        155720769 B/op   3639313 allocs/op
BenchmarkCheckBrokerStorageRemovalMap-12              14         408982641 ns/op        158881898 B/op   3639976 allocs/op

Because almost impossible to have 1000broker cluster, I would stay with the for loop.

Reply:
The important field here is the ns/op

Reply:
I added benchmark tests, and you and the team can also check. Before the merge those tests will be removed